### ISSUE TITLE
Date Objects in Pandas are mapped to NULLs

### ISSUE THREAD

**Issue by Admolly** [duckdb/duckdb#2155]
**What does happen?**
Essentially, when one has a column of datetime.date objects in a Pandas DataFrame, the relational mapper will create a VARCHAR column then map all the values to NULL.  Pandas does not have a specific type implemented for dates as it does with timestamps, so a column of datetime.date objects will convert to the generic dtype 'object'.  I am assuming this is why the mapper infers that the column should have type VARCHAR.

**What should happen?**
Since duckdb has a 'date' datatype, it would be nice if it could infer somehow that this type ought to be used instead of VARCHAR.  At a minimum, it would be nice if duckdb would instead store the string representation of the date object rather than converting it to NULL as in the example shown below.

**To Reproduce**
```
from datetime import date, datetime
import duckdb
import pandas as pd

df = pd.DataFrame.from_dict({
    "a":[date.today()],
    "b":[datetime.today()]
})

con = duckdb.connect()
rel = con.from_df(df)

print(rel.query('tmp', 'select * from tmp').fetchdf().to_markdown())

|    |   a | b                          |
|---:|----:|:---------------------------|
|  0 | nan | 2021-08-17 11:28:00.890197 |

print(df.applymap(type).to_markdown())

|    | a                       | b                                                  |
|---:|:------------------------|:---------------------------------------------------|
|  0 | <class 'datetime.date'> | <class 'pandas._libs.tslibs.timestamps.Timestamp'> |

print(df.dtypes.to_markdown())

|    | 0              |
|:---|:---------------|
| a  | object         |
| b  | datetime64[ns] |
```

Note that it's possible to convert the date object to a timestamp and normalize it, which will store it as a timestamp rather than a VARCHAR or date.

`df.a = pd.to_datetime(df.a).dt.normalize()`

**Environment (please complete the following information):**
 - OS: Windows 10
 - DuckDB Version: 0.2.8-dev919

**Before submitting**
- [X] Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?
- [ ] Have you tried this on the latest `master` branch? In case you cannot compile, you may find some binaries here: https://github.com/duckdb/duckdb/releases/tag/master-builds

**Comment by Mytherin** [comment]
Thanks for the report!

This should already be fixed on the master in #2092. The dates (and any other objects in a numpy.object array) should now be converted to strings. I get the following results now:

```py
>>> import duckdb
>>> import pandas as pd
>>> 
>>> df = pd.DataFrame.from_dict({
...     "a":[date.today()],
...     "b":[datetime.today()]
... })
>>> 
>>> con = duckdb.connect()
>>> rel = con.from_df(df)
>>> 
>>> print(rel.query('tmp', 'select * from tmp').fetchdf())
            a                          b
0  2021-08-17 2021-08-17 12:13:23.925783
```

Pandas doesn't really have any information about what is inside an object column, so we cannot know without looking at each individual value of the DataFrame. Right now we have opted to just convert everything in there to the varchar representation. You should be able to obtain a date column by using a cast:

```py
>>> print(rel.query('tmp', 'select a::date from tmp').fetchdf())
  CAST(a AS DATE)
0      2021-08-17
```

**Comment by Admolly** [comment]
Thanks for the info.  Probably no other way around it for the time being.  I suppose we can close the issue.
